src/main/java/com/couchbase/lite/replicator/PusherInternal.java

```diff
@@ -98,14 +115,48 @@ public void setCreateTarget(boolean createTarget) {
     }
 
     @Override
+    protected void start() {
+        // create single thread supportExecutor for push relication
+        initSupportExecutor();
+        super.start();
+    }
+
+    // create single thread supportExecutor for push replication
+    private void initSupportExecutor() {
+        if (supportExecutor == null || supportExecutor.isShutdown()) {
+            supportExecutor = Executors.newSingleThreadExecutor(new ThreadFactory() {
+                @Override
+                public Thread newThread(Runnable r) {
+                    String maskedRemote = remote.toExternalForm();
+                    maskedRemote = maskedRemote.replaceAll("://.*:.*@", "://---:---@");
+                    return new Thread(r, "CBLPusherSupportExecutor-" + maskedRemote);
+                }
+            });
+        }
+    }
+
+    // shutdown supportExecutor immediately
+    private void terminateSupportExecutor() {
+        if (supportExecutor != null && !supportExecutor.isShutdown()) {
+            Utils.shutdownAndAwaitTermination(supportExecutor, 0, 5);
+        }
+    }
+
+    @Override
     protected void stop() {
         if (stateMachine.isInState(ReplicationState.STOPPED))
             return;
 
-        Log.d(Log.TAG_SYNC, "%s STOPPING...", toString());
+        Log.d(TAG, "%s STOPPING...", toString());
 
         stopObserving();
 
+        // Awake thread if it is wait for pause
+        setPaused(false);
+
+        // shutdown supportExecutor immediately, does not add any more tasks.
+        terminateSupportExecutor();
```

Comment:
As you wrote, I believe race-condition does not cause the problem. Once executor is shutdown, any new tasks are not processed.